### PR TITLE
Big ugly bunch of changes

### LINKED ISSUE #44: Bug: USE WAREHOUSE
Due to the concurrent nature of terraform resource creation, the use of patterns like [USE WAREHOUSE](https://github.com/chanzuckerberg/terraform-provider-snowflake/blob/master/pkg/resources/warehouse.go#L64), USE DATABASE etc. is buggy. Each SQL statement needs to be self-contained since the DB connection is shared across all resources and hence different resources could be causing clashes on the USE statements.

This is especially true for statements like `select "name", "comment", "size" from table(result_scan(last_query_id()))` - there's no way to guarantee that the last query on the DB was the one from the current resource

### PR DESCRIPTION
**Pull request by sjauld**
Resolves #44, #43, #42 and #36 

Also a bunch of other stuff that I didn't create issues for:  here's (hopefully) a complete summary of what's included:

## New resources

* Database grant
* Resource monitor
* Schema
* Schema grant
* View
* View grant
* Warehouse grant

### Database

* Added the ability to create a database from a share
* Remove DB from state file if it no longer exists in the account

### Managed account

* Add a little sleep the first time a managed account is located since it is not immediately available to read

### Share

* Add the ability to add accounts to a share

### User

* Add the ability to force a password change upon first login

### Warehouse

* Add a whole pile of new properties
* Remove the use of the buggy "USE WAREHOUSE" sql call

## Other stuff

* Basically just some refactoring where I thought it was sensible

## Controversial stuff

Because we really need to execute all SQL statements in a single string, I have had to create fully qualified calls for all objects. This led me to create the ID convention of "<db_name>|<schema_name>|<object_name>|<privilege>". It's a bit of a hack but it was the best I could come up with at the time. We probably need to make sure we get something completely sensible in place here before releasing.

Sorry again for the huge PR - I kind of needed to get all this stuff up and running quickly so I just forged ahead!

### CODE REVIEW

File: pkg/resources/database.go [hunk 1]
```diff
@@ -27,10 +28,17 @@ var databaseSchema = map[string]*schema.Schema{
 		Optional: true,
 		Computed: true,
 	},
+	"from_share": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (sjauld) [reply pkg/resources/database.go]: Yes, definitely. I always get a little bit confused with maps - I'll have a read and try to fix it up.

File: pkg/resources/database_grant.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	dbName := data.Get("database_name").(string)
+	builder := snowflake.DatabaseGrant(dbName)
+	priv := data.Get("privilege").(string)
+
+	err := createGenericGrant(data, meta, builder)
+	if err != nil {
+		return err
+	}
+
+	// ID format is <db_name>|||<privilege>
```
[the remark this answers is not in the document]
Comment (sjauld) [reply pkg/resources/database_grant.go]: It's kind of similar to Route53 records in AWS (https://github.com/Viostream/terraform-provider-aws/blob/master/aws/resource_aws_route53_record.go#L370) - they use the underscore as a separator which works because no part of the identifier is allowed to contain one.

[the remark this answers is not in the document]
Comment (ryanking) [reply pkg/resources/database_grant.go]: Ok, just to be clear I am curious about whether the combination of db_name and privilege is uniq across a snowflake account.

**Comment by ryanking** [comment]
@sjauld if you want, I can take this PR over and get it merged

**Comment by sjauld** [comment]
Hi Ryan,

I'm happy for you to take it if you've got some time. I'm away for the
weekend!

Cheers,
Stuart
